### DISCUSSION TITLE
How is __typename exactly handled ?

### DISCUSSION THREAD

**Discussion by strblr** [urql-graphql/urql#1265] [Q&A] [answered]
I wanted to write something like this as an `update` after a mutation, where `thread` is an entity with `__typename` and `id` : 
```ts
cache.invalidate(newComment.thread);
```

But this doesn't work with TypeScript and Codegen because `__typename` is optional in the fragment type definition, and required as an input to `invalidate`. Currently, I don't explicitly ask for `__typename` in my fragments, but it seems like these are automatically added.

Which entity exactly does add them ? What are the guarantees behind this mechanism ?
And what would be the best **long-term** way to handle `__typename` :

- Explicitly listing it in the fragments
- Relying on whatever automatically adds them and set `nonOptionalTypename` to `true` in codegen ?

Thank you !

**Comment by JoviDeCroock** [comment, marked as answer]
In the documentation we talk about this https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#terminology

We need to be able to key an entity, we do this by means of `__typename:id`, this means that urql for either which cache will add the `__typename` field automatically for you, this will stay this way. This will stay this way, the main issue here seems to be only TypeScript so turning on the option to have it added to your types (mandatory) should be sufficient to make it work.

**Reply by strblr** [reply]
Is it the graphcache which does add them ?

The doc doesn't talk about it, it just says that it will log warnings if they're missing : 

> As we saw in the previous example, by default Graphcache will attempt to generate a key by combining the __typename of a piece of data with the id or _id fields, **if they're present**
> Graphcache will log a warning when these fields **weren't requested** as part of a query's selection set or aren't present in the data.

Regarding that last sentence, I don't request `__typename` but I don't have a warning either.

**Reply by JoviDeCroock** [reply]
> this means that urql for either which cache will add the __typename field automatically for you

@urql/core adds it by default to any operation that doesn't have it, the [`maskTypename`](https://formidable.com/open-source/urql/docs/api/core/#masktypename) bit elaborates on the specific function that does so.

**Reply by strblr** [reply]
Ok, so it's [`formatDocument`](https://formidable.com/open-source/urql/docs/api/core/#formatdocument). Thank you !
